ocaml/xapi/xapi_host.ml

```diff
@@ -596,10 +596,25 @@ let list_methods ~__context =
 let is_slave ~__context ~host = not (Pool_role.is_master ())
 
 let ask_host_if_it_is_a_slave ~__context ~host =
-  let local_fn = is_slave ~host in
-  Server_helpers.exec_with_subtask ~__context "host.ask_host_if_it_is_a_slave" (fun ~__context ->
-    (Message_forwarding.do_op_on_localsession_nolivecheck ~local_fn ~__context
-      ~host (fun session_id rpc -> Client.Client.Pool.is_slave rpc session_id host)))
+  let ask_if_slave_with_warn_log_when_slow ~__context =
+    let local_fn = is_slave ~host in
+    let timeout = 10. in
+    let task_name = Context.get_task_id __context |> Ref.string_of in
+    let log_host_slow_to_respond () =
+      D.warn "ask_host_if_it_is_a_slave: host taking a long time to respond - IP: %s; uuid: %s"
+        (Db.Host.get_address ~__context ~self:host) (Db.Host.get_uuid ~__context ~self:host)
+    in
+
```

Comment:
Perhaps rearm the scheduler with an exponential backoff so it is obvious if we are stuck for too long, and we don't flood the logs either:
```
let rec log_host_slow_to_respond timeout () =
....
Xapi_periodic_scheduler.add_to_queue task_name Xapi_periodic_scheduler.OneShot (log_host_slow_to_respond (timeout *. 2.))
in
```

Reply:
i tried to test with the following

```
let rec log_host_slow_to_respond timeout' () =
      D.warn "ask_host_if_it_is_a_slave: host taking a long time to respond - IP: %s; uuid: %s"
        (Db.Host.get_address ~__context ~self:host) (Db.Host.get_uuid ~__context ~self:host);
      Unix.sleep timeout';
      log_host_slow_to_respond (timeout') ();
    in

    Xapi_periodic_scheduler.add_to_queue task_name Xapi_periodic_scheduler.OneShot (float_of_int timeout) (log_host_slow_to_respond timeout);
    Unix.sleep 10;
    let res = Message_forwarding.do_op_on_localsession_nolivecheck ~local_fn ~__context ~host
      (fun session_id rpc -> Client.Client.Pool.is_slave rpc session_id host)
    in
    Xapi_periodic_scheduler.remove_from_queue task_name;
    res
  in
```
but it the logs keep on printing regardless; so it seems that once the scheduler has started executing the callback you have given it, removing the task from the queue will not kill the callback

Reply:
Don't call the function recursively, schedule a new Oneshot event instead.
Of course that once the callback starts it is not interruptable, this is not Lwt.

Reply:
Perhaps add a maximum interval when resetting the timer, e.g.: `min (timeout *. 2.) 300.`.